drivers/gles2/rasterizer_canvas_gles2.cpp

```diff
@@ -35,1403 +35,1532 @@
+		b.tex_pixel_size.y = 1.0;
 	}
 
-	if (storage->frame.clear_request) {
-		glClearColor(storage->frame.clear_request_color.r,
-				storage->frame.clear_request_color.g,
-				storage->frame.clear_request_color.b,
-				state.using_transparent_rt ? storage->frame.clear_request_color.a : 1.0);
-		glClear(GL_COLOR_BUFFER_BIT);
-		storage->frame.clear_request = false;
-	}
+	if (p_tile) {
+		if (texture) {
+			// default
+			b.tile_mode = BatchTex::TILE_NORMAL;
 
-	/*
-	if (storage->frame.current_rt) {
-		glBindFramebuffer(GL_FRAMEBUFFER, storage->frame.current_rt->fbo);
-		glColorMask(1, 1, 1, 1);
+			// no hardware support for non power of 2 tiling
+			if (!storage->config.support_npot_repeat_mipmap) {
+				if (next_power_of_2(texture->alloc_width) != (unsigned int)texture->alloc_width && next_power_of_2(texture->alloc_height) != (unsigned int)texture->alloc_height) {
+					b.tile_mode = BatchTex::TILE_FORCE_REPEAT;
+				}
+			}
+		} else {
+			// this should not happen?
+			b.tile_mode = BatchTex::TILE_OFF;
+		}
+	} else {
+		b.tile_mode = BatchTex::TILE_OFF;
 	}
-	*/
 
-	reset_canvas();
+	// push back
+	bdata.batch_textures.push_back(b);
 
-	glActiveTexture(GL_TEXTURE0);
-	glBindTexture(GL_TEXTURE_2D, storage->resources.white_tex);
+	return bdata.batch_textures.size() - 1;
+}
 
-	glVertexAttrib4f(VS::ARRAY_COLOR, 1, 1, 1, 1);
-	glDisableVertexAttribArray(VS::ARRAY_COLOR);
+void RasterizerCanvasGLES2::_batch_upload_buffers() {
 
-	// set up default uniforms
+	// noop?
+	if (!bdata.vertices.size())
+		return;
 
-	Transform canvas_transform;
+	glBindBuffer(GL_ARRAY_BUFFER, bdata.gl_vertex_buffer);
 
-	if (storage->frame.current_rt) {
+	// orphan the old (for now)
+	glBufferData(GL_ARRAY_BUFFER, 0, 0, GL_DYNAMIC_DRAW);
 
-		float csy = 1.0;
-		if (storage->frame.current_rt && storage->frame.current_rt->flags[RasterizerStorage::RENDER_TARGET_VFLIP]) {
-			csy = -1.0;
-		}
-		canvas_transform.translate(-(storage->frame.current_rt->width / 2.0f), -(storage->frame.current_rt->height / 2.0f), 0.0f);
-		canvas_transform.scale(Vector3(2.0f / storage->frame.current_rt->width, csy * -2.0f / storage->frame.current_rt->height, 1.0f));
+	if (!bdata.use_colored_vertices) {
+		glBufferData(GL_ARRAY_BUFFER, sizeof(BatchVertex) * bdata.vertices.size(), bdata.vertices.get_data(), GL_DYNAMIC_DRAW);
 	} else {
-		Vector2 ssize = OS::get_singleton()->get_window_size();
-		canvas_transform.translate(-(ssize.width / 2.0f), -(ssize.height / 2.0f), 0.0f);
-		canvas_transform.scale(Vector3(2.0f / ssize.width, -2.0f / ssize.height, 1.0f));
+		glBufferData(GL_ARRAY_BUFFER, sizeof(BatchVertexColored) * bdata.vertices_colored.size(), bdata.vertices_colored.get_data(), GL_DYNAMIC_DRAW);
 	}
 
-	state.uniforms.projection_matrix = canvas_transform;
-
-	state.uniforms.final_modulate = Color(1, 1, 1, 1);
-
-	state.uniforms.modelview_matrix = Transform2D();
-	state.uniforms.extra_matrix = Transform2D();
-
-	_set_uniforms();
-	_bind_quad_buffer();
+	// might not be necessary
```

Comment:
As a side note. When upgrading to 4.0, we will need to go through our buffer management with a fine tooth comb and be careful about when we bind and unbind buffers.

For now its safest to always unbind when finished. But on web, the cost of the unbind is non-trivial.

Reply:
Yes there are still quite a few needless state changes (less than in the original), I didn't attempt to optimize them out yet (whole other source of bugs). It may be worth running all the gl commands through a thin wrapper layer in 4.0 in order to help detect / deal with this.

Reply:
I agree, we might actually get a lot of gains by tracking state ourselves. I think we may even be able to simplify the test joining function as well.